File: rust/src/two_sat.rs
include!("scc.rs");

struct TwoSat {
    n: usize,
    scc: Scc,
}

#[allow(dead_code)]
impl TwoSat {
    fn new(n: usize) -> Self {
        Self {
            n,
            scc: Scc::new(2 * n),
        }
    }

    fn add_clause(&mut self, i: usize, f: bool, j: usize, g: bool) {
        self.scc.add_edge(
            2 * i + (if f { 0 } else { 1 }),
            2 * j + (if g { 1 } else { 0 }),
        );
        self.scc.add_edge(
            2 * j + (if g { 0 } else { 1 }),
            2 * i + (if f { 1 } else { 0 }),
        );
    }

    fn find_assignment(&self) -> Option<Vec<bool>> {
        let components = self.scc.build_components();

        let mut assignment = vec![false; self.n];
        for i in 0..assignment.len() {
            if components[2 * i] == components[2 * i + 1] {
                return None;
            }

            assignment[i] = components[2 * i] < components[2 * i + 1];
        }

        Some(assignment)
    }
}

File: java/TwoSat.java
class TwoSat {
  int n;
  Scc scc;

  TwoSat(int n) {
    this.n = n;
    scc = new Scc(2 * n);
  }

  void addClause(int i, boolean f, int j, boolean g) {
    scc.addEdge(2 * i + (f ? 0 : 1), 2 * j + (g ? 1 : 0));
    scc.addEdge(2 * j + (g ? 0 : 1), 2 * i + (f ? 1 : 0));
  }

  boolean[] findAssignment() {
    int[] components = scc.buildComponents();

    boolean[] assignment = new boolean[n];
    for (int i = 0; i < assignment.length; ++i) {
      if (components[2 * i] == components[2 * i + 1]) {
        return null;
      }

      assignment[i] = components[2 * i] < components[2 * i + 1];
    }

    return assignment;
  }
}
